lib/navigation/control/navigator.cpp

```diff
@@ -0,0 +1,119 @@
+#include "navigator.hpp"
+
+#include <cstdint>
+
+namespace hyped::navigation {
+
+Navigator::Navigator(core::ILogger &logger, const core::ITimeSource &time)
+    : logger_(logger),
+      time_(time),
+      keyence_preprocessor_(logger),
+      accelerometer_preprocessor_(logger, time),
+      accelerometer_trajectory_estimator_(time),
+      crosschecker_(logger, time),
+      running_means_filter_(logger, time)
+{
+  // TODOLater: implement, add log and timesource so far
+  // TODOLater: instantiate everything?
+}
+
+std::optional<core::Trajectory> Navigator::currentTrajectory()
+{
+  core::Float mean_encoder_value = 0;
+  for (std::size_t i = 0; i < core::kNumEncoders; ++i) {
+    mean_encoder_value += static_cast<core::Float>(previous_encoder_reading_.at(i));
+  }
+  mean_encoder_value /= core::kNumEncoders;
+
+  core::Float mean_keyence_value = 0;
+  for (std::size_t i = 0; i < core::kNumKeyence; ++i) {
+    mean_keyence_value += static_cast<core::Float>(previous_keyence_reading_.at(i));
+  }
+  mean_keyence_value /= core::kNumKeyence;
+
+  const SensorChecks check_trajectory = crosschecker_.checkTrajectoryAgreement(
+    trajectory_.displacement, mean_encoder_value, mean_keyence_value);
+
+  if (check_trajectory == SensorChecks::kUnacceptable) {
+    logger_.log(core::LogLevel::kFatal,
+                "Navigation sensors are in disagreement. Unable to accurately determine "
+                "trajectory.");
+    return std::nullopt;
+  }
+
+  return trajectory_;
+}
+
+void Navigator::keyenceUpdate(const core::KeyenceData &keyence_data)
+{
+  // Check keyence strictly increasing
+  if (keyence_data.at(0) < previous_keyence_reading_.at(0)) {
+    logger_.log(core::LogLevel::kFatal, "Keyence data is decreasing");
+  }
```

Comment:
do we plan to add anything other than a logger to indicate this failure?

Reply:
such as?

Reply:
core::Result?

Reply:
What does that do differently than a logged failure?

Reply:
logging doesn't trigger any behaviour, if we need to react to this event then we must return something indicating an error

Reply:
Ok. I'll change the return types of the main navigation interfaces then

Reply:
Ok. I'll change the return types of the main navigation interfaces then